enhancements/microshift/dummy-gateway-interface-ovn.md

```diff
@@ -0,0 +1,476 @@
+
+The general flow is:
+
+1. The packet enters host via eth0.
+2. In the host, the packet is DNATed to the service's clusterIP by the following iptables rule.
+
+```none
+-A OVN-KUBE-NODEPORT -p tcp -m addrtype --dst-type LOCAL -m tcp --dport 30847 -j DNAT --to-destination 10.43.117.48:80
+```
+
+3. The packet is forwarded to br-ex bridge.
+
+The rest of the steps is the same as the normal local gateway mode.
+
+### Workflow Description
+
+#### Deploying
+
+There will be no change to the MicroShift deployment procedure.
+
+#### Upgrading
+
+The microshift-ovs-init.service will be responsible for the OVS bridge change.
+As we will use a dummy interface for br-ex bridge, the `gatewayInterface` flag
+in `/etc/microshift/ovn.yaml` will be deprecated. Also, as the configuration of
+OVS will not affect host physical interfaces, there is no need to allow user to
+config the OVS bridges. The flag `disableOVSInit` can also be deprecated.
+
+#### Configuring
+
+There is no need to introduce any new configuration flag for this change. Unless
+we want to allow user to specify the IP address used by the br-ex interface.
+
+#### Deploying Applications
+
+Applications are deployed as usual.
+
+#### Variation [optional]
+
+N/A
+
+### API Extensions
+
+N/A
+
+### Implementation Details/Notes/Constraints
+
+On the OVN-Kubernetes side, we need to change the local gateway mode:
+
+1. Allow using an external gateway bridge without uplink port
+2. Add static routes for all host network with the ovn-k internal masquerade
+   next-hop.
+
+On the MicroShift side, we will need to make the following change:
+
+1. We need to update the microshift-ovs-init.service to 
+   * Create br-ex bridge, and assign a static IP to it.
+   * Since we don't need NetworkManager to manage the DHCP client. We can remove
+   those logic from the configure-ovs.sh.
+   * For upgrading, we also need to remove the NetworkManager connections
+   created by configure-ovs.sh in previous version.
+
+### Risks and Mitigations
+
+The OVN-Kubernetes configuration described in this document is not used by any
+other known platforms. Therefore there is no CI with this setup in
+OVN-Kubernetes upstream. In order to minimize the risk of regression, we need to
+improve the OVN-Kubernetes CI to cover this setup.
+
+### Drawbacks
+
+The egress features that rely on sending packets to wire without passing kernel
+will not work. As there is no uplink port to forward those packets to. So
+egressGW and egressIP will not work. A possible solution is to forward those
+traffic to the host kernel, then adding iptables rules to steer the traffic to
+wire.
+
+With this change, the ovn node annotations such
+`k8s.ovn.org/node-primary-ifaddr` and `k8s.ovn.org/l3-gateway-config` will use
+the local IP of br-ex as the node IP, instead of the IP of host's primary
+interface. Therefore the following features that need to read these node
+annotation will be affected. The egress rules of Egress firewall contains the
```

Comment:
which egress firewall rules contain the node IP? The policies we create to bypass egress firewall if trying to talk to node ip? If so, this is a bug. We should have acls to bypass egress firewall for talking to any ip on the host.

Reply:
I am not familiar with egress firewall code. I found we add nodeIP to the rules at https://github.com/ovn-org/ovn-kubernetes/blob/65642570543ffc4f82e53af25789c3bb28bac41d/go-controller/pkg/ovn/egressfirewall.go#L702.

Reply:
ah right I added this code. This is for using a node selector in the egress firewall to match IPs. I think this is a bug, we should match on all node IPs. Can you file a bug for this one too?

Reply:
https://issues.redhat.com/browse/OCPBUGS-13665 was created.